# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
#load python packages
import os
import pandas as pd
from collections import Counter
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
from sklearn.preprocessing import StandardScaler

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
df=pd.read_csv('step3_df_output.csv',index_col=0)
print(df['Name'])
df.head()

0                         Alyeska Resort
1                    Eaglecrest Ski Area
2                       Hilltop Ski Area
3                       Arizona Snowbowl
4                    Sunrise Park Resort
                     ...                
325                 Meadowlark Ski Lodge
326            Sleeping Giant Ski Resort
327                     Snow King Resort
328    Snowy Range Ski & Recreation Area
329                  White Pine Ski Area
Name: Name, Length: 330, dtype: object


Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        NaN          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             NaN         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             NaN         0  
4          78.0              104.0            80.0         0  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
df=pd.get_dummies(df, columns=['state'])

In [4]:
df=df.fillna(0)

In [5]:
df.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  state_Rhode Island  \
0        0.0          0          2     2       0  ...                   0   
1        0.0          0          0     0       0  ...                   0   
2        0.0          0          0     0       1  ...                   0   
3        0.0          1          0     2       2  ...                   0   
4        0.0          0          1     2       3  ...                   0   

   state_South Dakota  state_Tennessee  state_Utah  state_Vermont  \
0                   0                0           0              0   
1                   0                0           0              0   
2                   0                0           0              0   
3                   0                0           0              0   
4                   0                0           0              0   

   state_Virginia  state_Washington  state_West Virginia  state_Wisconsin  \
0               0                 0                    0                0   
1               0                 0                    0                0   
2               0                 0                    0                0   
3               0                 0                    0                0   
4               0                 0                    0                0   

   state_Wyoming  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 61 columns]

In [6]:
df2 = df.loc[:, ~df.columns.str.startswith('state')]

In [7]:
df.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  state_Rhode Island  \
0        0.0          0          2     2       0  ...                   0   
1        0.0          0          0     0       0  ...                   0   
2        0.0          0          0     0       1  ...                   0   
3        0.0          1          0     2       2  ...                   0   
4        0.0          0          1     2       3  ...                   0   

   state_South Dakota  state_Tennessee  state_Utah  state_Vermont  \
0                   0                0           0              0   
1                   0                0           0              0   
2                   0                0           0              0   
3                   0                0           0              0   
4                   0                0           0              0   

   state_Virginia  state_Washington  state_West Virginia  state_Wisconsin  \
0               0                 0                    0                0   
1               0                 0                    0                0   
2               0                 0                    0                0   
3               0                 0                    0                0   
4               0                 0                    0                0   

   state_Wyoming  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 61 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [8]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
x = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler().fit(x)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(x) 

In [9]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
x2 = df2.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler2 = StandardScaler().fit(x2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled2=scaler2.transform(x2) 

In [33]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing
df3 = df.loc[:, ~df.columns.str.startswith('state')]
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
x3 = df3.drop(['Name','AdultWeekend','summit_elev','base_elev'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler3 = StandardScaler().fit(x3)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled3=scaler3.transform(x3) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [11]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [12]:
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train2, X_test2, y_train2, y_test2 = train_test_split(X_scaled2, y, test_size=0.25, random_state=1)

In [13]:
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train3, X_test3, y_train3, y_test3 = train_test_split(X_scaled3, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [14]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [15]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [16]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import mean_squared_error
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

In [17]:
explained_variance_score(y_test, y_pred)

0.9313360598539226

In [18]:
explained_variance_score( y_pred,y_test)

0.919518969829173

In [19]:
mean_absolute_error(y_test, y_pred)

5.147133387191184

In [20]:
mean_absolute_error(y_pred,y_test)

5.147133387191184

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [21]:
print(lm.intercept_)
print(lm.coef_)

64.1250269849115
[-2.08476976e+01  6.51962014e+00  1.90280345e+01 -1.67329535e+12
 -1.64501324e+11 -1.94744096e+12 -6.56919390e+12 -3.92140007e+12
 -4.83846908e+12 -5.42387497e+12 -6.15484207e+12  1.73282903e+13
 -1.28491211e+00 -1.94404602e-01 -2.97851562e-01 -1.96777344e-01
  3.14620972e-01  2.61230469e-01  3.87390137e-01 -3.36669922e-01
  1.91933594e+01 -1.00585938e-01  8.43750000e-01 -1.09765625e+00
 -1.02664190e+11 -8.39530354e+10 -2.64042217e+11 -2.69818164e+11
 -1.32132961e+11 -2.02483046e+11 -1.18364994e+11 -8.39530354e+10
 -1.02664190e+11 -1.76180669e+11 -5.94541853e+10 -1.94167334e+11
 -3.06243395e+11 -2.18017845e+11 -8.39530354e+10 -2.02483046e+11
 -1.18364994e+11 -2.32332143e+11 -8.39530354e+10 -1.76180669e+11
 -3.24503719e+11 -1.44521551e+11 -1.32132961e+11 -1.85421236e+11
 -2.51965731e+11 -5.94541853e+10 -8.39530354e+10 -5.94541853e+10
 -2.10419406e+11 -2.25312554e+11 -1.18364994e+11 -1.85421236e+11
 -1.18364994e+11 -2.32332143e+11 -1.66363256e+11]


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [22]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
cofdf=pd.DataFrame(abs(lm.coef_), x.columns, columns=['Coefficient'])
cofdf.iloc[:10]

Coefficient
summit_elev    2.084770e+01
vertical_drop  6.519620e+00
base_elev      1.902803e+01
trams          1.673295e+12
fastEight      1.645013e+11
fastSixes      1.947441e+12
fastQuads      6.569194e+12
quad           3.921400e+12
triple         4.838469e+12
double         5.423875e+12

In [23]:
cofdf.sort_values('Coefficient').tail(10)

Coefficient
state_Michigan  3.062434e+11
state_New York  3.245037e+11
trams           1.673295e+12
fastSixes       1.947441e+12
quad            3.921400e+12
triple          4.838469e+12
double          5.423875e+12
surface         6.154842e+12
fastQuads       6.569194e+12
total_chairs    1.732829e+13

In [24]:
cofdf

Coefficient
summit_elev           2.084770e+01
vertical_drop         6.519620e+00
base_elev             1.902803e+01
trams                 1.673295e+12
fastEight             1.645013e+11
fastSixes             1.947441e+12
fastQuads             6.569194e+12
quad                  3.921400e+12
triple                4.838469e+12
double                5.423875e+12
surface               6.154842e+12
total_chairs          1.732829e+13
Runs                  1.284912e+00
TerrainParks          1.944046e-01
LongestRun_mi         2.978516e-01
SkiableTerrain_ac     1.967773e-01
Snow Making_ac        3.146210e-01
daysOpenLastYear      2.612305e-01
yearsOpen             3.873901e-01
averageSnowfall       3.366699e-01
AdultWeekday          1.919336e+01
projectedDaysOpen     1.005859e-01
NightSkiing_ac        8.437500e-01
clusters              1.097656e+00
state_Alaska          1.026642e+11
state_Arizona         8.395304e+10
state_California      2.640422e+11
state_Colorado        2.698182e+11
state_Connecticut     1.321330e+11
state_Idaho           2.024830e+11
state_Illinois        1.183650e+11
state_Indiana         8.395304e+10
state_Iowa            1.026642e+11
state_Maine           1.761807e+11
state_Maryland        5.945419e+10
state_Massachusetts   1.941673e+11
state_Michigan        3.062434e+11
state_Minnesota       2.180178e+11
state_Missouri        8.395304e+10
state_Montana         2.024830e+11
state_Nevada          1.183650e+11
state_New Hampshire   2.323321e+11
state_New Jersey      8.395304e+10
state_New Mexico      1.761807e+11
state_New York        3.245037e+11
state_North Carolina  1.445216e+11
state_Ohio            1.321330e+11
state_Oregon          1.854212e+11
state_Pennsylvania    2.519657e+11
state_Rhode Island    5.945419e+10
state_South Dakota    8.395304e+10
state_Tennessee       5.945419e+10
state_Utah            2.104194e+11
state_Vermont         2.253126e+11
state_Virginia        1.183650e+11
state_Washington      1.854212e+11
state_West Virginia   1.183650e+11
state_Wisconsin       2.323321e+11
state_Wyoming         1.663633e+11

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [25]:
model2 = lm.fit(X_train2,y_train2)

In [26]:
print(explained_variance_score(y_test2, y_pred))
print(explained_variance_score( y_pred,y_test2))
print(mean_absolute_error(y_test2, y_pred))
print(mean_absolute_error(y_pred,y_test2))

0.9313360598539226
0.919518969829173
5.147133387191184
5.147133387191184


In [27]:
cofdf2=pd.DataFrame(abs(lm.coef_), x2.columns, columns=['Coefficient'])
cofdf2.iloc[:10]

Coefficient
summit_elev      13.667005
vertical_drop     4.895438
base_elev        10.486503
trams             0.302032
fastEight         0.088317
fastSixes         0.425019
fastQuads         0.700191
quad              1.559395
triple            1.447859
double            0.377383

In [28]:
print(lm.intercept_)
print(lm.coef_)

64.06180964776385
[-1.36670054e+01  4.89543804e+00  1.04865034e+01 -3.02031585e-01
 -8.83168515e-02  4.25018788e-01  7.00191389e-01  1.55939455e+00
  1.44785872e+00  3.77383252e-01 -1.30931122e+00  6.93442971e-01
 -1.48289775e+00 -2.04288799e-01 -1.65870322e-01 -4.04754006e-01
 -1.59614668e-03  1.01605478e+00  2.90103177e-01 -1.58251646e+00
  1.99983708e+01 -3.76790155e-01  5.16848706e-01 -9.45255710e-01]


In [29]:
X_train2

array([[-1.58669848e-01, -9.14414942e-01,  4.04832918e-02, ...,
        -1.04716512e+00, -6.09611897e-01, -7.80868809e-02],
       [-2.55187912e-01,  1.47351277e+00, -7.53117745e-01, ...,
         5.20812283e-01, -6.09611897e-01, -7.80868809e-02],
       [-9.76392337e-01, -9.67245201e-01, -8.75210213e-01, ...,
        -9.90324405e-16, -3.31010103e-01, -7.80868809e-02],
       ...,
       [-9.69689694e-01, -9.93660331e-01, -8.27015818e-01, ...,
        -6.98725699e-01, -2.34571021e-01, -7.80868809e-02],
       [ 1.05853018e+00,  2.61464637e+00,  8.43723208e-01, ...,
         6.44428248e+00,  2.60502419e+00,  1.53245504e+00],
       [ 1.97330695e+00, -5.64515333e-02,  2.38273089e+00, ...,
         5.55656226e-01, -6.09611897e-01, -1.68862880e+00]])

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

In [34]:
model3 = lm.fit(X_train3,y_train3)

In [35]:
print(explained_variance_score(y_test3, y_pred))
print(explained_variance_score( y_pred,y_test3))
print(mean_absolute_error(y_test3, y_pred))
print(mean_absolute_error(y_pred,y_test3))

0.9313360598539226
0.919518969829173
5.147133387191184
5.147133387191184


In [36]:
x3.shape

(330, 22)

#### Model 3

In [37]:
print(lm.intercept_)
print(lm.coef_)

64.05787081240787
[ 1.31142086 -0.24685943  0.08025693  0.41751786  0.80387844  1.61140832
  1.40087092  0.37498631 -1.28850004  0.74412825 -1.62267074 -0.10779556
 -0.15520885 -0.47980774 -0.14075529  0.88996284  0.31323762 -2.04559897
 19.9533297  -0.12970099  0.65388921 -0.78055032]


In [38]:
cofdf=pd.DataFrame(abs(lm.coef_), x3.columns, columns=['Coefficient'])
cofdf.iloc[:10]

Coefficient
vertical_drop     1.311421
trams             0.246859
fastEight         0.080257
fastSixes         0.417518
fastQuads         0.803878
quad              1.611408
triple            1.400871
double            0.374986
surface           1.288500
total_chairs      0.744128

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.93 | 5.147133387191184 |-|
| Model 2. | 0.93|5.147133387191184 |'state'|
| Model 3. | 0.93 | 5.147133387191184 |'state','summit_elev','base_elev'|

Model Selection:

I would go with model two.  Model 1 gives too much information and creates noise that cause issues when trying to utilize the data later on.  Model three takes out elevation, which is a determining factor when finding pricing on ski resorts(including the drop).  I noticed the explained variance and mean absolute errors remained the same and will wait to see if there is more differences in information.